v3/mathjax3-ts/util/LinkedList.ts

```diff
@@ -0,0 +1,356 @@
+     *
+     * @return{DataClass[]}  The list converted to an array
+     */
+    public toArray() {
+        return Array.from(this) as DataClass[];
+    }
+
+    /*
+     *  Used for sorting and merging lists (Overridden by subclasses)
+     *
+     * @param{DataClass} a   The first item to compare
+     * @param{DataClass} b   The second item to compare
+     * @return{boolean}      True if a is before b, false otherwise
+     */
+    public isBefore(a: DataClass, b: DataClass) {
+        return (a < b);
+    }
+
+    /*
+     * Push items on the end of the list
+     *
+     * @param{DataClass[]} args   The list of data items to be pushed
+     * @return{LinkedList}        The LinkedList object (for chaining)
+     */
+    public push(...args: DataClass[]) {
+        for (const data of args) {
+            let item = new ListItem<DataClass>(data);
+            item.next = this.list;
+            item.prev = this.list.prev;
+            this.list.prev = item;
+            item.prev.next = item;
+        }
+        return this;
+    }
+
+    /*
+     * Pop the end item off the list and return its data
+     *
+     * @return{DataClass}  The data from the last item in the list
+     */
+    public pop(): DataClass {
+        let item = this.list.prev;
+        if (item.data === END) {
+            return null;
+        }
+        this.list.prev = item.prev;
+        item.prev.next = this.list;
+        item.next = item.prev = null;
+        return item.data as DataClass;
+    }
+
+    /*
+     * Push items at the head of the list
+     *
+     * @param{DataClass[]} args   The list of data items to inserted
+     * @return{LinkedList}        The LinkedList object (for chaining)
+     */
+    public unshift(...args: DataClass[]) {
+        for (const data of args.slice(0).reverse()) {
+            let item = new ListItem<DataClass>(data);
+            item.next = this.list.next;
+            item.prev = this.list;
+            this.list.next = item;
+            item.next.prev = item;
+        }
+        return this;
+    }
+
+    /*
+     * Remove an item from the head of the list and return its data
+     *
+     * @return{DataClass}  The data from the first item in the list
+     */
+    public shift(): DataClass {
+        let item = this.list.next;
+        if (item.data === END) {
+            return null;
+        }
+        this.list.next = item.next;
+        item.next.prev = this.list;
+        item.next = item.prev = null;
+        return item.data as DataClass;
+    }
+
+    /*
+     * Empty the list
+     *
+     * @return{LinkedList}  The LinkedList object (for chaining)
+     */
+    public clear() {
+        this.list.next = this.list.prev = this.list;
+        return this;
+    }
+
+    /*
+     * Make the list iterable and return the data from the items in the list
+     *
+     * @return{{next: Function}}  The object containing the iterator's next() function
+     */
+    public [Symbol.iterator]() {
```

Comment:
This is what makes an object iterable (i.e., can be used in for-of statements).

See [this description](http://2ality.com/2015/02/es6-iteration.html#implementing-iterables) for more details.